include/envoy/http/codec.h

```diff
@@ -148,11 +148,45 @@ class ConnectionCallbacks {
 };
 
 /**
- * A list of options that can be specified when creating a codec.
+ * HTTP/2 codec settings
  */
-class CodecOptions {
-public:
-  static const uint64_t NoCompression = 0x1;
+struct Http2Settings {
+  // TODO(jwfang): support other HTTP/2 settings
+  uint32_t hpack_table_size_{DEFAULT_HPACK_TABLE_SIZE};
+  uint32_t max_concurrent_streams_{DEFAULT_MAX_CONCURRENT_STREAMS};
+  uint32_t initial_window_size_{DEFAULT_INITIAL_WINDOW_SIZE};
+  uint32_t initial_connection_window_size_{DEFAULT_INITIAL_CONNECTION_WINDOW_SIZE};
+
+  // disable HPACK compression
+  static const uint32_t MIN_HPACK_TABLE_SIZE = 0;
+  // initial value from HTTP/2 spec, same as NGHTTP2_DEFAULT_HEADER_TABLE_SIZE from nghttp2
+  static const uint32_t DEFAULT_HPACK_TABLE_SIZE = (1 << 12);
+  // no maximum from HTTP/2 spec, use unsigned 32-bit maximum
+  static const uint32_t MAX_HPACK_TABLE_SIZE = (1UL << 32) - 1;
+
+  // prevent creation of new streams from peer
+  static const uint32_t MIN_MAX_CONCURRENT_STREAMS = 0;
+  // defaults to maximum, same as nghttp2
+  static const uint32_t DEFAULT_MAX_CONCURRENT_STREAMS = (1U << 31) - 1;
+  // no maximum from HTTP/2 spec, total streams is unsigned 32-bit maximum
+  // one-side (client/server) is half that, and we need to exclude stream 0
+  // same as NGHTTP2_INITIAL_MAX_CONCURRENT_STREAMS from nghttp2
+  static const uint32_t MAX_MAX_CONCURRENT_STREAMS = (1U << 31) - 1;
+
+  // initial value from HTTP/2 spec, same as NGHTTP2_INITIAL_WINDOW_SIZE from nghttp2
+  // NOTE: we only support increase window size now, so this is also the minimum
+  // TODO(jwfang): make this 0 to support decrease window size
+  static const uint32_t MIN_INITIAL_WINDOW_SIZE = (1 << 16) - 1;
+  // initial stream-level value from HTTP/2 spec, same as NGHTTP2_INITIAL_WINDOW_SIZE from nghttp2
+  static const uint32_t DEFAULT_INITIAL_WINDOW_SIZE = (1 << 16) - 1;
```

Comment:
I've noticed that you've changed the default from current `256MB` to `64KB` resulting in 4096x smaller window size.

As previously noted, `256MB` effectively disables flow control, so I definitely agree that we should lower it, but such drastic change will result in different traffic characteristics, and therefor should be committed separately, otherwise we're risking reverting this commit as a whole.

Maybe let's leave it as `256MB` in this PR and let's change the default to `1MB` in subsequent PR?

@htuch any thoughts?